src/notebooks/execution/cellExecutionMessageHandler.ts

```diff
@@ -0,0 +1,574 @@
+            }
+        };
+        request.onReply = (msg) => {
+            // Cell has been deleted or the like.
+            if (this.cell.document.isClosed) {
+                request.dispose();
+            }
+            this.handleReply(msg);
+        };
+        request.onStdin = this.handleInputRequest.bind(this);
+        request.done.finally(() => this.endCellExecution());
+    }
+
+    private clearLastUsedStreamOutput() {
+        this.lastUsedStreamOutput = undefined;
+    }
+    /**
+     * Assume we run cell A
+     * Now run cell B, and this will update cell A.
+     * The way it works is, the request object created for cell A will get a message saying update your output.
+     * Cell A has completed, hence there's no execution task, we should create one or re-use an existing one.
+     * Creating one results in side effects such as execution order getting reset and timers starting.
+     * Hence where possible re-use an existing cell execution task associated with this document.
+     */
+    private createTemporaryTask() {
+        if (this.cell.document.isClosed) {
+            return;
+        }
+        // If we have an active task, use that instead of creating a new task.
+        const existingTask = activeNotebookCellExecution.get(this.cell.notebook);
+        if (existingTask) {
+            return existingTask;
+        }
+
+        // Create a temporary task.
+        this.previousResultsToRestore = { ...(this.cell.executionSummary || {}) };
+        this.temporaryExecution = CellExecutionCreator.getOrCreate(this.cell, this.controller);
+        this.temporaryExecution?.start();
+        if (this.previousResultsToRestore?.executionOrder && this.execution) {
+            this.execution.executionOrder = this.previousResultsToRestore.executionOrder;
+        }
+        return this.temporaryExecution;
+    }
+    private endTemporaryTask() {
+        if (this.previousResultsToRestore?.executionOrder && this.execution) {
+            this.execution.executionOrder = this.previousResultsToRestore.executionOrder;
+        }
+        if (this.previousResultsToRestore && this.temporaryExecution) {
+            if (this.previousResultsToRestore.executionOrder) {
+                this.temporaryExecution.executionOrder = this.previousResultsToRestore.executionOrder;
+            }
+            this.temporaryExecution.end(
+                this.previousResultsToRestore.success,
+                this.previousResultsToRestore.timing?.endTime
+            );
+        } else {
+            // Undefined for not success or failure
+            this.temporaryExecution?.end(undefined);
+        }
+        this.previousResultsToRestore = undefined;
+        this.temporaryExecution = undefined;
+    }
+    private handleIOPub(msg: KernelMessage.IIOPubMessage) {
+        // eslint-disable-next-line @typescript-eslint/no-require-imports
+        const jupyterLab = require('@jupyterlab/services') as typeof import('@jupyterlab/services');
```

Comment:
Any benefit in requiring here vs importing from the top? I’m guessing this loads heavy stuff?

Reply:
Yes, thats correct, and we don't want webpack to inline these imports either.